Tensorboard logging and EOS weights

liaojh1998 opened the pull request:
This branch is used to test out the new changes (RNN/Transformers).
- Added tensorboard logging for all weights.
- Added EOS weights for loss to focus on stop tokens better.

The code is untested as my code is currently getting `RuntimeError: CUDA error: misaligned address`.

liaojh1998 committed:
Correcting environments for UNAST on conda

liaojh1998 committed:
Correcting environments for UNAST on conda

liaojh1998 committed:
Adding loss weights for unbalanced classes

liaojh1998 committed:
Logging output to tensorboard

liaojh1998 committed:
Removed unnecessary print statements

liaojh1998 committed:
Fixup from merging with decoder branch

liaojh1998 committed:
Checking for nans and separating batch sizes for train and eval

liaojh1998 committed:
Reporting more consise losses and testing out transformers

Lucaskabela commented on src/rnn_no_discrim.json:
```diff
@@ -1,26 +1,29 @@
+    "teacher_decay_end": 200,
     "grad_clip": 1.0,
-    "batch_size": 16,
+    "train_batch_size": 4,
```
Might be able to try a larger batch size no with the:
`with torch.no_grad():`
in inference!  I can reliably do 16

Lucaskabela commented on src/train.py:
```diff
@@ -19,16 +20,17 @@
 class BatchGetter():
     def __init__(self, args, supervised_dataset, unsupervised_dataset, full_dataset):
-        self.batch_size = args.batch_size
+        self.batch_size = args.train_batch_size
```
Good insight on splitting this into train/eval batch size!

Lucaskabela commented on src/train.py:
```diff
@@ -482,6 +618,21 @@ def train(args):
+    if WRITER:
+        ex_mel = torch.tensor(ex["mel"]).unsqueeze(0)
+        ex_text = torch.LongTensor(ex["text"]).unsqueeze(0)
+        text_pred = model.asr(None, ex_mel.to(DEVICE), infer=True).squeeze(dim=0).detach().cpu().numpy()
```
need to pass lengths here!

Lucaskabela commented on src/train.py:
```diff
@@ -482,6 +618,21 @@ def train(args):
+        text_pred = model.asr(None, ex_mel.to(DEVICE), infer=True).squeeze(dim=0).detach().cpu().numpy()
+        speech_pred, speech_stop_pred = model.tts(ex_text.to(DEVICE), None, infer=True)
+        speech_pred = speech_pred.squeeze(dim=0).detach().cpu().numpy()
+        WRITER.add_text(f"{name}/text_gold", sequence_to_text(ex["text"]), step)
```
We should trim this to length if it is not all ready, same with speech prediction!

liaojh1998 committed:
Passed length to asr() and tts(), trimmed model example output during inference

liaojh1998 committed:
Added schedulers for transformer training

liaojh1998 committed:
Fixed outputs of speech transformer output for matching old code

liaojh1998 commented:
Rebased on top of latest `mel_loss/bill`.

Added:
- Schedulers for changing learning rate for transformers.

Code is tested using train and eval batch size of 8.

Some notes on the `RuntimeError`:
- The error only occurs in `crossmodel_step()`, and specifically, `model.cm_speech_in(mel, mel_len)` every time when trying to do inference on the encoded speech sequence for text.
- The error only occurs on accessing `stop_lens[stop_mask] = i` in text inference. Try printing `stop_lens` or printing `choice` there will result in this error as well.

liaojh1998 commented:
`RuntimeError` example log, but happens during `evaluate()` call as well.
![image](https://user-images.githubusercontent.com/22307142/116083261-0c66c680-a662-11eb-8067-1ed31cf428ab.png)

liaojh1998 committed:
Fixed continue training portion, changed flatten to squeeze

Lucaskabela committed:
Merge branch 'loss-fns-evaluate-gpu/Jay' of https://github.com/Lucaskabela/UNAST into loss-fns-evaluate-gpu/Jay

liaojh1998 committed:
Fixed discriminator step logging

liaojh1998 committed:
Added discriminator run

liaojh1998 committed:
[WIP] Added argument to save every some number of epochs

Commit:
fix d for embedding on rnn

liaojh1998 committed:
Fixed saving every epoch using repeated arguments, fixed scheduler steps when using MultistepLR

Commit:
added lstm discrim

Commit:
add testing discriminator jsons

Commit:
update hidden layer passing for disciminator

Commit:
fixed up rnn and transformer encoder outptus for disc

Lucaskabela committed:
1. Pack in RNN 2. Mask in Trans 3. Sqrt

Lucaskabela committed:
Merge branch 'loss-fns-evaluate-gpu/Jay' of https://github.com/Lucaskabela/UNAST into loss-fns-evaluate-gpu/Jay

Lucaskabela committed:
Small fixes

Lucaskabela committed:
Merge branch 'loss-fns-evaluate-gpu/Jay' into d_hotfix/bill

Lucaskabela committed:
1. One hidden for LSTMD 2. Add lengths

Lucaskabela committed:
Fix to cpu

Lucaskabela committed:
Missed one name

liaojh1998 commented:
The RNN optimizations here looks good. Thanks for working on this Lucas!

Lucaskabela committed:
Fix double down projection issue

liaojh1998 committed:
Shuffled text and speech examples for discriminator batch

Lucaskabela committed:
Clean up things a bit

liaojh1998 committed:
Added all runs and configs

liaojh1998 committed:
Binarized discrim, log discrim output, and continue training when kicked from condor

liaojh1998 committed:
Added continuing run configs

liaojh1998 committed:
Merge pull request #17 from Lucaskabela/d_hotfix/bill

Discriminator hotfix from Bill
